Task/763 payments implementation

taoeffect opened the pull request:
Closes almost, but not entirely: #763 #206

taoeffect committed:
fix some copy errors (#873)

taoeffect committed:
first draft payments implementation with adjusted distribution

taoeffect committed:
some work on supporting late payments

taoeffect committed:
much improved late payments handling + updated dependencies!

taoeffect committed:
fix rollup-plugin-eslint bug

sandrina-p commented on frontend/model/contracts/group.js:
```diff
@@ -54,22 +59,92 @@ export function createInvite (
-    payments: {},
-    frozenDistribution: null,
-    frozenMincome: null
+    globalUsername: 'TODO: this? e.g. groupincome:greg / namecoin:bob / ens:alice',
```
I don't understand this comment (that actually is a string).

taoeffect replied:
Yeah, this isn't being used at the moment... I'll change this to `''` and add a `// TODO:` after it. It's something to remind me to consider for the future.

sandrina-p commented on frontend/model/contracts/group.js:
```diff
@@ -118,31 +193,58 @@ DefineContract({
+    // adjusted version of groupIncomeDistribution, used by the payments system
+    groupIncomeAdjustedDistribution (state, getters) {
+      const monthstamp = currentMonthstamp()
+      return groupIncomeDistribution({ state, getters, monthstamp, adjusted: true })
```
another way:
```js
return getters.groupIncomeAdjustedDistributionForMonth(monthstamp)
```

sandrina-p commented on frontend/model/contracts/group.js:
```diff
@@ -450,6 +575,10 @@ DefineContract({
         if (Object.keys(invite.responses).length === invite.quantity) {
           invite.status = INVITE_STATUS.USED
         }
+        // TODO: ensure `meta.username` is unique for the lifetime of the username
```
Is there any way where `meta.username` wouldn't be the same? A user won't be able to change their username, right?

taoeffect replied:
Yeah I don't think they'll be able to change their username... this comment is just to help us think about how accounts and identities are registered and handled in the future.

sandrina-p commented on frontend/model/contracts/voting/proposals.js:
```diff
@@ -122,24 +122,27 @@ const proposals = {
       sbp('gi.contracts/group/removeMember/process', message, state)
+      // TODO: this await usage is a violation... we shouldn't
+      //       have any async calls / side-effects inside /process
       await sbp('gi.sideEffects/group/removeMember', {
```
I said that when this was implemented. `removeMember` doesn't even need to be an `async function`. IMO both can be reverted to sync functions..

sandrina-p commented on frontend/views/containers/proposals/RuleRemoveMember.vue:
```diff
@@ -17,7 +17,7 @@ proposal-template(
       name='changeReason'
       ref='purpose'
-      :placeholder='L("The reason why I\' propositiong this change is...")'
+      :placeholder='L("The reason I\'m proposing this change is...")'
```
Actually, in Figma there isn't any placeholder. We could just remove this in the 3 proposals.

sandrina-p commented on frontend/views/utils/currencies.js:
```diff
@@ -1,29 +1,49 @@
-  const integerPart = /^(.+)\.0+$/.exec(fixed)
-  return integerPart ? integerPart[1] : fixed
+// https://github.com/okTurtles/group-income-simple/issues/813#issuecomment-593680834
+// round all accounting to DECIMALS_MAX decimal places max to avoid consensus
+// issues that can arise due to different floating point values
+// at extreme precisions. If this becomes inadequate, instead of increasing
+// this value, switch to a different currency base, e.g. from BTC to mBTC.
```
I think this comment could be removed. That information is already on the mentioned github issue.

taoeffect replied:
Without this comment someone looking at this code would be confused. We can't expect developers to memorize everything in our issues.

sandrina-p replied:
Maybe I wasn't clear enough. I totally agree with what you are saying. My suggestion was just to leave a comment:
`// [1 line explanation]. See more at #813`

taoeffect replied:
Eh... can we leave it as is?

sandrina-p commented on frontend/views/containers/payments/PaymentsList.vue:
```diff
@@ -143,6 +146,74 @@ export default {
+      // first we need to check if there are any late payments that need to be added to the paymentsTodo list
+      if (pastMonth) {
+        const previousDistribution = pastMonth.lastAdjustedDistribution
+        // this creates a date at the end of the previous month
```
A UI component (Vue, React, etc...) main responsibility is to map data to the UI, avoiding creating any extra complex logic, leaving that part to the "store / backend part", in our case the Vuex getters.

In my opinion, this logic could be a getter `this.paymentsLateDistribution` and then the UI only concacts both:

```js
return this.paymentsLateDistribution.concat(this.paymentsDistribution)
```

Another advantage of this approach is we can create a unit test to make sure this works as expected (since with Cypress it will be _a big challenge_ to simulate late payments.)

taoeffect replied:
Maybe. It makes sense to make it a getter only if it's used more than once. I'll think about this more when I get to writing tests. 👍

sandrina-p commented on frontend/views/containers/payments/PaymentsList.vue:
```diff
@@ -143,6 +146,74 @@ export default {
+            //         current month's distribution.
+            // Let D = the amount we're pledging this month
+            // Let E = the amount still unpaid for the previous month's distribution,
+            //         calculated as: C > 0 ? A : A + D - B
```
I understand your reason to "minify" the variable names. What if we take a middle step, where we can reduce the nr of characters while still understanding the meaning of the variables:
```
A -> prevOwe
B -> sent
C -> curOwe
D -> pldg
C -> unpaid
```

Clarity is important, especially when dealing with complex logic.

taoeffect replied:
Is it OK if we keep it as-is? If it really bugs you feel free to modify it in a future PR and adjust the spacing in the comments so that everything lines up pretty!

sandrina-p replied:
Your decision.

sandrina-p commented on frontend/model/contracts/group.js:
```diff
@@ -203,65 +307,64 @@ DefineContract({
         toUser: string,
         amount: number,
-        currency: string,
+        currency: string, // must be one of the keys in currencies.js (e.g. USD, EUR, etc.)
```
Can't we have a checker like `oneOf(['USD', 'EUR'])` or `oneOf(Object.keys(currencies))`?

taoeffect replied:
In the next update I plan to change this to:

```js
currencyFromTo: tupleOf(string, string), // must be one of the keys in currencies.js (e.g. USD, EUR, etc.)
```

The reason why I'm keeping it as `string, string` is because `currencies.js` will be constantly updated. Doing `oneOf(Object.keys(currencies))` would break clients that are out of date when one of the clients is updated with new currencies.

sandrina-p reviewed (commented):
Here's my initial review. I'll be waiting for the unit tests to see this stuff in action!

taoeffect committed:
add a new way of handling sideEffects in contracts

taoeffect committed:
semi-functional refactoring and redesign of payments

taoeffect commented:
Summary of recent changes:

### Refactored and changed how Payments.vue works

- Removed as much `v-if` usage as possible and replaced it with components that represent rows in a table, improving the clarity of the code and maybe giving a slight boost in performance. This included creating a component for each type of row that can be displayed in the table: `PaymentRowReceived`, `PaymentRowSent`, `PaymentRowTodo`, and `PaymentRowRecord`. **These files also have a lot of excess CSS that was copy/pasted and will be removed in a future commit.**
- There are now two different type of PaymentLists: `PaymentsList.vue`, which is the main one, and `RecordPaymentsList.vue`, which represents the one in the modal when payments are being recorded. This was done to again remove `v-if` usage and to make it easier to reason about the code and both read and modify it.
- Computed properties were created for `Payments.vue` which now return the list of payments for each tab, and support searching/filtering the list based on all 3 properties: username, displayName, and amount.
- The file `PaymentsListMenu.vue` was moved into `historical/` and its contents were moved into the appropriate row components, again to remove `v-if` usage.

### Remaining issues

- One fairly significant issue remaining involves the following scenario: users `A`, `B` and `C` are in a group. `A` is pledging and `B` and `C` are receiving mincome. `A`'s pledge is enough to cover some but not all of `B` and `C`'s needs. `A` makes a single complete payment to `B`. In this case the payment is successfully sent and reflected in the UI, however the algorithm re-runs the distribution afresh, and since `B`'s needs were still not completely satisfied, it now again recommends another payment to `B`, and a slightly smaller payment to `C` than it recommended the first time. This is not how it's supposed to behave. Fixing this will be a significant part of the work done in the next update to this PR.
- As mentioned above, when creating the row components for the table, a lot of the CSS was simply copied and pasted from a single other file, and some of it doesn't belong in many of the `.vue` files in this PR. A lot of it needs to be removed, and some of it might need to be consolidated together outside of the components and in the main `.scss` files. I might need your help with this @sandrina-p @pieer (as far as suggestions on where to put it).
- @mmbotelho's design calls for marking payments as completed immediately. This isn't how the payments code was designed to work. Instead it was designed for a message to be sent indicating a new and pending payment, for which a followup message was expected to mark it as completed. Margarida's manual payments mode instead calls for payments to be immediately marked as completed when they're created, and then possibly marked as "uncompleted" ("not received"). The code needs to be updated to handle this.

taoeffect committed:
more or less functioning payments implementation

taoeffect committed:
removed historical/group.js

taoeffect committed:
fix tests

taoeffect commented:
### Changes

Rudimentary payments now work!

- You can see your TODO list, and pay all or pay individuals, and change the amount you send
- Sending partial payments, and seeing them in the list
- Marking payments as not received
- Changing your income details (amounts, or even the income type)
- Maybe even works with late payments (though I haven't tried that yet, only implemented it)

Miscellaneous changes:

- Forgot to mention in the previous update that there's a new way of doing sideEffects in the contract; search for `pushSideEffect` in the codebase to see how it's used in `group.js` and `proposals.js`
- Updated a lot of dependencies to fix security issues

### What doesn't work

- Figuring out what to do when a payment is marked as not received (no proper followup to this) — @mmbotelho @dotmacro please play with it and see if you can offer suggestions
- Proper display of payment due dates and late payment dates
- Sidebar overview of payments (the stats on the righthand side)
- Proper display of currencies based on the group's currency setting
- Conversion of currencies or handling currency change proposals
- Payments list pagination
- Reset button in RecordPayments modal, and other functionality like the memo field (didn't hook that up, needs to be done, but should be straightforward now)
- And probably other stuff

### What needs to be done next

- Someone needs to go through and clean up all of the CSS — deduplicate the all of the copy/pasted junk in `PaymentRow*.vue` files and DRY it all so that it's all stored in one of the `.scss` files and not repeated
- Create mixins and remove no longer used computed properties/methods from the `PaymentRow*.vue` files (and any other payment related files)
- Implement and use the `prevMonthDueDate` computed property for late payments in `PaymentRowTodo.vue`
- Delete `humanDate.js` and move the `humanDate` function into `time.js`
- Delete outdated comments in the markup
- Cleanup and DRY the markup, CSS, and any JS, but especially the CSS
- Remove the `Math.random()` stuff in the `PaymentRow*.vue` files (mostly related to payment date stuff) and replace it with the actual data
- Correctly implement the `PaymentDetails.vue` file
- Correctly implement the `MonthOverview.vue` file
- Cancelling payments? Not sure how that flow is supposed to work, wasn't clear from Figma (cc @mmbotelho)
- Sorting and reverse sorting the payments list by clicking the tabs
- Maybe make it so that payments can be marked as completed immediately when `paymentType === PAYMENT_TYPE_MANUAL` (see TODO in `/payment` in `group.js`), although this one isn't that important, it works well enough as it is now without it (two messages sent, one `/payment` followed by a `/paymentUpdate`)
- Making the payment date information use the time given by the server instead of the system clock (so no `new Date()` usage, but creating dates based on an API call, see #531 and the `/time` API defined on the backend)
- Fixing a bug where if we switched from sending payments to receiving payments, we suddenly lose the "Sent" tab
- Making the `ESC` key clear the search field
- Implement anything I forgot to mention from the "what doesn't work" section above
- **Error handling!!** I did not add any proper error handling or error displaying for sending payments, or for the input forms. Vuelidate validations need to be added where appropriate as well.
- **Tests!!** I did not add a single test for to this, and it will be challenging to test some things like late payments since that requires time manipulation of some kind

The above can be extracted and grouped into new issues to be closed. @sandrina-p and @pieer I'm wondering if you can take some of these on? (Just make sure to create an issue and self-assign if one doesn't exist). It will free me to focus on PR reviews and much overdue blog posts!

mmbotelho commented:
> Figuring out what to do when a payment is marked as not received (no proper followup to this) — @mmbotelho @dotmacro please play with it and see if you can offer suggestions

[This flow is specified on Figma](https://www.figma.com/file/mxGadAHfkWH6qApebQvcdN/Group-Income-2.0?node-id=9518%3A21) - Read the orange notes there for more info! If you stillhave questions, let me know!

> Cancelling payments? Not sure how that flow is supposed to work, wasn't clear from Figma (cc @mmbotelho)

For manual payments, the payment is no longer “done”, therefore it returns to its original state in the todo page. It’s basically an undo.

pieer commented on frontend/views/containers/payments/PaymentRowReceived.vue:
```diff
@@ -0,0 +1,323 @@
+
+    &:last-child {
+      width: 40%;
+      min-width: 9,375rem;
```
9.375rem ?

taoeffect replied:
Fixed! (not my code!)

pieer commented on frontend/views/containers/payments/PaymentRowReceived.vue:
```diff
@@ -0,0 +1,323 @@
+  }
+}
+
+// TODO: check if we need generic rule for this (keep until we know)
```
That or we can use the class .has-background-warning

taoeffect replied:
No idea, didn't write this comment. Maybe you or @sandrina-p can fix this in a future PR? Fixing the CSS is a big part of the followup to this (see [comment](https://github.com/okTurtles/group-income-simple/pull/900#issuecomment-622661198))

sandrina-p commented on frontend/views/containers/payments/RecordPayment.vue:
```diff
@@ -68,64 +66,103 @@ modal-base-template(ref='modal' :fullscreen='true' class='has-background' v-if='
+              txid: '' + Math.random(),
+              status: PAYMENT_PENDING,
+              paymentType: PAYMENT_TYPE_MANUAL
+              // TOOD: add memo if user added a note
```
What's "memo" ?

taoeffect replied:
it's a key on the `/payment` message that can be added. Using it for the user-note on payment seems perfect.

sandrina-p commented on frontend/views/containers/payments/PaymentsList.vue:
```diff
@@ -1,113 +1,44 @@
       required: true
     },
-    payments: {
+    paymentsList: {
```
A validation is needed at `paymentsType` prop to make sure the string matches one of the components name.

sandrina-p commented on frontend/views/containers/proposals/ProposalVoteOptions.vue:
```diff
@@ -102,25 +102,26 @@ export default {
         const proposalHash = this.proposalHash
-        const isOneVoteToPass = oneVoteToPass(proposalHash)
-        const payload = {}
+        var passPayload = {}
```
let

sandrina-p commented on frontend/model/contracts/group.js:
```diff
@@ -441,28 +568,51 @@ DefineContract({
-      process ({ data, meta }, { state }) {
-        Vue.delete(state.profiles, data.member)
+      process ({ data, meta }, { state, getters }) {
+        state.profiles[data.member].status = PROFILE_STATUS.REMOVED
+        state.profiles[data.member].departedDate = meta.createdDate
```
Why the need to change `removeMember` logic in a PR related with payments stuff?

taoeffect replied:
This was done because of the way that the payments distributions were calculated in a previous commit, which relied on keeping track of who was an active member during each month, in order to calculate the distribution for that month.

In a later commit I stopped doing it that way, but did not revert this part. I decided to keep it so that who said what in chat histories could be preserved. Margarida also has designed a profile card for members who are no longer active in the group, and this makes it easier to implement that.

sandrina-p commented on frontend/model/contracts/group.js:
```diff
@@ -441,28 +568,51 @@ DefineContract({
-      validate: objectOf({
-        groupId: string
+      validate: objectMaybeOf({
+        reason: string
```
Everywhere `removeOurselves` is called, I didn't see `reason` to be called in any one of the places. In what case should a reason be given? And how can the user add it in the UI?

taoeffect replied:
They might be able to give an explanation for why they're leaving. It's optional though so I don't think it's a big deal if we leave it here? Harder to add back after the protocol is fully defined.

pieer reviewed (changes requested):
Great Job Greg! Such a long PR, but great things happening here!
I feel that PaymentRowRecord, PaymentRowSent and PaymentRowTodo could be drier. 70% of the code is the same in those 3 files.

sandrina-p commented on frontend/views/containers/payments/PaymentRowTodo.vue:
```diff
@@ -0,0 +1,330 @@
+      return humanDate(datems, { month: 'short', day: 'numeric' })
+    },
+    // TODO: make multiple payments
+    async cancelPayment () {
+      try {
```
I tried this and it didn't work. The error:

> TypeValidatorError: invalid "undefined" value type; string type expected

sandrina-p commented on frontend/views/containers/payments/RecordPayment.vue:
```diff
@@ -11,10 +11,8 @@ modal-base-template(ref='modal' :fullscreen='true' class='has-background' v-if='
-          :payments='paymentsDistribution'
-          paymentsType='edit'
+        record-payments-list(
+          :paymentsList='ephemeral.payments'
```
Scouting rule: Add css class`.textarea` to `textarea` on line 39. The styles are missing there.

sandrina-p commented:
I found 3 bugs: maybe related to the other you described:
- Once a member marks a payment as "not received", the other member cannot re-record the payment again.
- I mark that I need 20$. User2 sends me the money. Then I update my income details saying I need 30$. User2 is asked to send me 30$, instead of 10$.
- I mark that I need 20$. User2 sends me the money. Then I update my income details saying I need 10$. User2 is asked to send me 10$ again, but it should be asked only on the next month because he already sent me $20 before.

taoeffect committed:
ensure a sideEffectStack per contract instance

dotmacro commented:
A bug, possibly related to the others:
- User1 has already made payments up to their full pledge amount. User2 adjusts their income details saying they need more. User1 is asked to make an additional payment i.e. to pay more than their maximum pledge.

taoeffect committed:
review feedback

taoeffect commented:
OK, I just pushed a commit that addresses the review feedback but not the bugs that you guys found above.

I have edited the top comment so that it will not close the issues when merged. I am hoping that @sandrina-p you can take over [the remaining TODOs](https://github.com/okTurtles/group-income-simple/pull/900#issuecomment-622661198) from here!

sandrina-p reviewed (dismissed):
A lot still needs to be done, but I'm okay to merge this and do it in another PR @taoeffect .

sandrina-p commented:
> I feel that PaymentRowRecord, PaymentRowSent and PaymentRowTodo could be drier. 70% of the code is the same in those 3 files. 

I agree with @pieer. This is a perfect example where named slots would work perfectly for a single component `<PaymentRow>`.

taoeffect commented:
> I agree with @pieer. This is a perfect example where named slots would work perfectly for a single component <PaymentRow>.

If `PaymentRow` can be implemented without `v-if`s cluttering things up, then I'm OK with that. I don't see how that can be done though. The reason I split up the original component into multiple ones was to avoid the `v-if` usage, which both degrades performance and decreases readability.

sandrina-p commented:
@taoeffect I have a question about the data structure of each key under `group.payments`

```js
{
  "21XWnNKKz4XWYV4KX2qVCuVxDa94JJDrSik7hWmRGXkv6uq81A": {
    "data": {
      "toUser": "user2-1041",
      "amount": 3.02,
      "currencyFromTo": [
        "USD",
        "USD"
      ],
      "exchangeRate": 1,
      "txid": "0.993821200796108",
      "status": "not-received",
      "paymentType": "manual",
      "completedDate": "2020-05-06T16:54:54.731Z"
    },
    "meta": {
      "createdDate": "2020-05-06T16:54:54.689Z",
      "username": "user1-1041",
      "identityContractID": "21XWnNVQNrta9BUe2PQjyzwFmpRNVBBYG1uXUKFCxySvggnxUU"
    },
    "history": [
      [
        "2020-05-06T16:54:54.689Z",
        {
          "toUser": "user2-1041",
          "amount": 3.02,
          "currencyFromTo": [
            "USD",
            "USD"
          ],
          "exchangeRate": 1,
          "txid": "0.993821200796108",
          "status": "completed",
          "paymentType": "manual",
          "completedDate": "2020-05-06T16:54:54.731Z"
        }
      ],
      [
        "2020-05-06T16:54:54.731Z",
        {
          "status": "completed"
        }
      ],
      [
        "2020-05-06T17:16:08.202Z",
        {
          "status": "not-received"
        }
      ]
    ]
  }
}
```

What's the diff between `history[0]` and `data`? Why should it be stored twice?

--- EDIT ---

Oh gotcha, the data is the actual current state. History is all the data and modifications like git commits. Forget my question 😶

taoeffect commented:
> Oh gotcha, the data is the actual current state. History is all the data and modifications like git commits. Forget my question 😶

@sandrina-p It's actually a good question because it made me realize something: there's no need to store the actual data in the history, it can instead store the hash as a reference to the message that modified the payment. I'll send another commit to do that... will let you know when that's done.

taoeffect committed:
payment history now consists of [date, hash] pairs

taoeffect commented:
> will let you know when that's done.

Done!

sandrina-p commented on frontend/model/contracts/group.js:
```diff
@@ -417,7 +417,7 @@ DefineContract({
           throw new Errors.GIErrorIgnoreAndBanIfGroup('paymentUpdate from bad user!')
         }
-        payment.history.push([meta.createdDate, data.updatedProperties])
+        payment.history.push([meta.createdDate, hash])
```
Dumb question: How do I access a message's data by its hash?

For example, when we access `history[0][1]`, it will retrieve the hash, but then how would we access the actual data?

taoeffect replied:
@sandrina-p not a dumb question at all!

Each message is stored by hash via `addMessageToDB()`  in `state.js`, which calls `'gi.db/log/addEntry'`, which in turn is defined in `shared/domains/gi/db.js`, and uses the hash as the key into IndexedDB.

So you would call `'gi.db/log/getEntry'` using the hash to retrieve it.

sandrina-p replied:
Gotcha! In what situations/UI will we need to access a payment history?

taoeffect replied:
@sandrina-p it's possible something like that might need to be done when showing the details of a payment.

Similarly, we will need to use that API to show the complete payment history, or anything that involves digging up history of anything (e.g. chat messages, etc.), since we cannot store everything in Vuex memory all the time.

sandrina-p replied:
It makes sense. Thanks!

taoeffect committed:
Merge branch 'master' into task/763-payments-implementation

taoeffect committed:
Merge branch 'master' into task/763-payments-implementation

taoeffect committed:
add strings updates
